modules/system/console/CreateMigration.php

```diff
@@ -199,6 +203,120 @@ protected function prepareVars(): array
+            }
+            if ($type === 'fields') {
+                $fieldList = $fields[$type];
+            } else {
+                $fieldList = $fields[$type]['fields'];
+            }
+
+            foreach ($fieldList as $field => $config) {
+                if (str_contains($field, '@')) {
+                    list($field, $context) = explode('@', $field);
+                }
+
+                $type = $config['type'] ?? 'text';
+
+                if (str_starts_with($field, '_')
+                    or $field === $model->getKeyName()
+                    or str_contains($field, '[')
+                    or in_array($type, ['fileupload','relation','relationmanager','section','hint'])
+                    or in_array($field, $model->purgeable ?? [])
+                    or $model->getRelationType($field)
+                ) {
+                    continue;
+                }
+
+                $vars['fields'][$field] = $this->mapFieldType($field, $config);
+            }
+        }
+
+        foreach ($model->getRelationDefinitions() as $relationType => $definitions) {
+            if (in_array($relationType, ['belongsTo', 'hasOne'])) {
+                foreach (array_keys($definitions) as $relation) {
+                    $vars['fields'][$relation . '_id'] = [
+                        'type' => 'foreignId',
+                        'index' => true,
+                        'required' => true,
+                    ];
+                }
+            }
+        }
+
+        if ($model->methodExists('getSortOrderColumn')) {
+            $field = $model->getSortOrderColumn();
+            $vars['fields'][$field] = [
+                'type' => 'unsignedinteger',
+                'required' => false,
+                'index' => true,
+            ];
+        }
+
+        $vars['primaryKey'] = $model->getKeyName();
+        $vars['jsonable'] = $model->getJsonable();
+        $vars['timestamps'] = $model->timestamps;
+
+        if ($morphable = $model->morphTo) {
+            $vars['morphable'] = array_keys($morphable);
+        }
+
+        return $vars;
+    }
+
+    protected function mapFieldType($name, $fieldConfig)
+    {
+        switch ($fieldConfig['type'] ?? 'text') {
+            case 'checkbox':
+            case 'switch':
+                $dbType = 'unsignedinteger';
+                break;
+            case 'number':
+            case 'range':
+                $dbType = 'integer';
+                break;
+            case 'datepicker':
+                $dbType = 'timestamp';
+                break;
+            case 'markdown':
+            case 'textarea':
+                $dbType = 'text';
+                break;
+            default:
+                $dbType = 'string';
+        }
+        $required = $fieldConfig['required'] ?? false;
+
+        return [
+            'type' => $dbType,
+            'required' => $required,
+            'index' => in_array($name, ["slug"]) or str_ends_with($name, "_id"),
+        ];
```

Comment:
Can we move the mapping logic from fields.yaml to table structure into a separate helper class that can be more easily unit tested? Additionally, is there any overlap here with the Builder plugin?

Reply:
> Can we move the mapping logic from fields.yaml to table structure into a separate helper class that can be more easily unit tested?

Any suggestion ?

> Additionally, is there any overlap here with the Builder plugin?

Not really, from what I could see, this is done at a different level in Builder.

Reply:
> Can we move the mapping logic from fields.yaml to table structure into a separate helper class that can be more easily unit tested?

Done!